File: tools/bench/answers/go_sonnet/s2/09_topwords.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	sentence := "the cat sat on the mat with the cat"
	words := strings.Fields(sentence)
	counts := make(map[string]int)
	order := []string{}
	for _, w := range words {
		if counts[w] == 0 {
			order = append(order, w)
		}
		counts[w]++
	}
	top1, top2 := "", ""
	cnt1, cnt2 := 0, 0
	for _, w := range order {
		c := counts[w]
		if c > cnt1 {
			top2, cnt2 = top1, cnt1
			top1, cnt1 = w, c
		} else if c > cnt2 {
			top2, cnt2 = w, c
		}
	}
	fmt.Println(top1, cnt1)
	fmt.Println(top2, cnt2)
}

File: tools/bench/answers/rust_sonnet/s1/09_topwords.rs
fn main() {
    let sentence = "the cat sat on the mat with the cat";
    let mut words: Vec<&str> = Vec::new();
    let mut counts: Vec<(&str, usize)> = Vec::new();
    for word in sentence.split_whitespace() {
        let mut found = false;
        for entry in counts.iter_mut() {
            if entry.0 == word {
                entry.1 += 1;
                found = true;
                break;
            }
        }
        if !found {
            counts.push((word, 1));
        }
        let _ = words;
    }
    counts.sort_by(|a, b| b.1.cmp(&a.1));
    for i in 0..2 {
        println!("{} {}", counts[i].0, counts[i].1);
    }
}
